### PR TITLE
Content Migration to Server

### PR DESCRIPTION
**Pull request by ditadewindita**
Moved interest system and access control functionalities from client side to store data to `Server`.

All new data additions/modifications log its status to the server log and multiple clients will now access the same state of chat data. Slight modifications were made in `Chat` to compensate for the new data access method.

Transaction log feature has been updated to take care of loading interest system and access control data. This load takes place in `Server` and utilizes the server's `Controller` class.

**TL;DR** - All data has been moved to server, huzzah!

### CODE REVIEW

File: src/codeu/chat/client/core/Controller.java [hunk 1]
```diff
@@ -38,6 +42,264 @@ public Controller(ConnectionSource source) {
...
+    } catch (Exception ex){
+      System.out.println("ERROR: Exception during call on server. Check log for details.");
+      LOG.error(ex, "Exception during call on server.");
+    }
+
+    return response;
+  }
+
+  @Override
+  public Integer toggleCreatorBit(Uuid convo, Uuid user, Boolean flag){
```
Comment (kevmo314) [inline src/codeu/chat/client/core/Controller.java]: The above three methods are very similar, can you condense the code to be something like:

```
@Override
public Integer toggleCreatorBit(Uuid convo, Uuid user, Boolean flag) {
  return toggleAccessControlBit(convo, user, flag, NetworkCode.TOGGLE_CREATOR_BIT_REQUEST, NetworkCode.TOGGLE_CREATOR_BIT_RESPONSE);
}

private Integer toggleAccessControlBit(Uuid convo, Uuid user, Boolean flag, Integer requestBit, Integer responseBit) {
   ...
  Serializers.INTEGER.write(connection.out(), requestBit);
  ...
  if(Serializers.INTEGER.read(connection.in()) == responseBit) {  ... }
}
```

and reuse that method for all three?

File: src/codeu/chat/server/Controller.java [hunk 2]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
 
   @Override
+  public Integer toggleRemovedBit(Uuid convo, Uuid user){
+    final ConversationHeader foundConvo = model.conversationById().first(convo);
+    final User foundUser = model.userById().first(user);
+
+    Integer result = 0;
+
+    if(foundConvo != null && foundUser != null) {
+      Integer access = foundConvo.accessControls.computeIfAbsent(user, newAccess -> 0);
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java]: `accessControls.getOrDefault(user, 0);`

File: src/codeu/chat/server/Controller.java [hunk 3]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+      Integer newAccess;
+
+      // Flag is set to true if the user is removed from a conversation.
+      // Flag stays true once it's set.
+      newAccess = access | ConversationHeader.REMOVED;
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java]: Just do this along with the definition:

`Integer newAccess = access | ConversationHeader.REMOVED;`

File: src/codeu/chat/server/Controller.java [hunk 4]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+              "toggleCreatorBit fail (user.id=%s conversation.id=%s)",
+              foundUser.id,
+              foundConvo.id);
+    }
+
+    return result;
+  }
+
+  @Override
+  public Integer toggleOwnerBit(Uuid convo, Uuid user, Boolean flag){
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java:131]: Same thing with these methods, can they be simplified and call a `private Integer toggleAccessBit(Uuid convo, Uuid user, Boolean flag, Integer accessBit);`?

File: src/codeu/chat/server/Controller.java [hunk 5]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+
+  @Override
+  public Integer updateUsersUnseenMessagesCount(Uuid user, Uuid convo, Integer count){
+    final User foundUser = model.userById().first(user);
+    final ConversationHeader foundConvo = model.conversationById().first(convo);
+
+    Integer result = 0;
+
+    if(foundUser != null && foundConvo != null){
+      Integer currentCount = (foundConvo.unseenMessages.get(foundUser) == null) ? 0 : foundConvo.unseenMessages.get(foundUser);
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java]: `unseenMessages.getOrDefault(foundUser, 0);`

File: src/codeu/chat/server/Controller.java [hunk 6]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+      Integer setCount = (currentCount + count < 0) ? 0 : currentCount + count;
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java]: `Math.max(0, currentCount + count);`

File: src/codeu/chat/server/Controller.java [hunk 7]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+              time.inMs());
+    } else {
+
+      LOG.info(
+              "updateUsersLastStatusUpdate fail (user.id=%s time=%s)",
+              foundUser.id,
+              time.inMs());
+    }
+
+    return update;
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java:260]: You can avoid needing to define `update` if you do `return foundUser != null ? foundUser.lastStatusUpdate : null;`.

Alternatively, move the `return` into the if/else and return the appropriate values.

File: src/codeu/chat/server/Controller.java [hunk 8]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+              time.inMs());
+    } else {
+      LOG.info(
+              "newUpdatedConversation fail (user.id=%s conversation.id=%s time=%s)",
+              foundUser.id,
+              foundConversation.id,
+              time.inMs());
+    }
+
+    return map;
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java:287]: Same for `map` as above.

File: src/codeu/chat/server/Controller.java [hunk 9]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+              foundUser.id,
+              followedUser.id);
+    } else {
+      LOG.info(
+              "newUserInterest fail - user/followed user doesn't exist (user.id=%s user.id=%s)",
+              foundUser.id,
+              followedUser.id);
+    }
+
+    return foundUser.userInterests;
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java]: `foundUser` might be null here.

File: src/codeu/chat/server/Controller.java [hunk 10]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+              foundUser.id,
+              followedUser.id);
+    } else {
+      LOG.info(
+              "removeUserInterest fail - user/followed user doesn't exist (user.id=%s user.id=%s)",
+              foundUser.id,
+              followedUser.id);
+    }
+
+    return foundUser.userInterests;
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java]: Same, might be null.

File: src/codeu/chat/server/Controller.java [hunk 11]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+              foundUser.id,
+              foundConvo.id);
+    } else {
+      LOG.info(
+              "newConversationInterest fail - user/conversation does not exist (user.id=%s conversation.id=%s)",
+              foundUser.id,
+              foundConvo.id);
+    }
+
+    return foundUser.conversationInterests;
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java]: Same.

File: src/codeu/chat/server/Controller.java [hunk 12]
```diff
@@ -56,6 +59,326 @@ public ConversationHeader newConversation(String title, Uuid owner) {
...
+              foundUser.id,
+              foundConvo.id);
+    } else {
+      LOG.info(
+              "removeConversationInterest fail - user/conversation doesn't exist (user.id=%s conversation.id=%s)",
+              foundUser.id,
+              foundConvo.id);
+    }
+
+    return foundUser.conversationInterests;
```
Comment (kevmo314) [inline src/codeu/chat/server/Controller.java]: Same.

File: src/codeu/chat/util/store/Store.java [hunk 13]
```diff
@@ -74,6 +74,12 @@ public VALUE first(KEY key) {
   }
 
   @Override
+  public VALUE last(KEY key) {
```
Comment (kevmo314) [inline src/codeu/chat/util/store/Store.java]: Is this method actually used anywhere? Regardless, because maps can only have one key, this seems like it would be the same as `first()`?

File: src/codeu/chat/client/core/Controller.java (hunk 1 above)
Comment (ditadewindita) [reply src/codeu/chat/client/core/Controller.java]: Condensing it does seem better, thank you! I made the generic `toggleAccessControlBit` method private within the Controller.

File: src/codeu/chat/server/Controller.java (hunk 2 above)
Comment (ditadewindita) [reply src/codeu/chat/server/Controller.java]: Thank you for this, I didn't know `getOrDefault` existed. So is `computeIfAbsent` only for collections?

File: src/codeu/chat/server/Controller.java (hunk 3 above)
Comment (ditadewindita) [reply src/codeu/chat/server/Controller.java]: Simplified what I didn't fix when copy pasting, thank you :)

File: src/codeu/chat/server/Controller.java (hunk 4 above)
Comment (ditadewindita) [reply src/codeu/chat/server/Controller.java:131]: Hmm..I looked over the code to see to implement this, and I think that the `toggleAccessBit` method would just get too long? I also have some specific log statements in each specific bit toggle method currently.

File: src/codeu/chat/server/Controller.java (hunk 7 above)
Comment (ditadewindita) [reply src/codeu/chat/server/Controller.java:260]: I kept it the way the base code did it - having a variable that was null and changing it if necessary. Seems like it fits better stylistically with everything else.

File: src/codeu/chat/server/Controller.java (hunk 9 above)
Comment (ditadewindita) [reply src/codeu/chat/server/Controller.java]: Yup, noticed this while doing the access controls and forgot to fix it. Thanks! All nulls have been checked for this server's Controller class :)

File: src/codeu/chat/util/store/Store.java (hunk 13 above)
Comment (ditadewindita) [reply src/codeu/chat/util/store/Store.java]: Did this method during testing and must've forgotten to remove this one, too. Everything would work with a first() since its only one key, hehe.

File: src/codeu/chat/server/Controller.java (hunk 2 above)
Comment (kevmo314) [reply src/codeu/chat/server/Controller.java]: `computeIfAbsent` is also for maps but inserts the computed value into the map. In this case you don't need to flood the map with 0's so just using a default is sufficient.

File: src/codeu/chat/server/Controller.java (hunk 4 above)
Comment (kevmo314) [reply src/codeu/chat/server/Controller.java:131]: Ah, I see. This is fine then.

**Review by kevmo314 — APPROVED** [review]
